temporal-sdk/src/main/java/io/temporal/client/ListWorkflowExecutionIterator.java

```diff
@@ -0,0 +1,133 @@
+
+  private final @Nonnull String query;
+  private final @Nonnull String namespace;
+  private final int pageSize;
+  private final @Nonnull GenericWorkflowClient genericClient;
+  private ListWorkflowExecutionsResponse activeResponse;
+  private int nextActiveResponseIndex;
+  private CompletableFuture<ListWorkflowExecutionsResponse> nextResponse;
+
+  public ListWorkflowExecutionIterator(
+      @Nonnull String query,
+      @Nonnull String namespace,
+      int pageSize,
+      @Nonnull GenericWorkflowClient genericClient) {
+    this.query = Objects.requireNonNull(query, "query");
+    this.namespace = Objects.requireNonNull(namespace, "namespace");
+    this.pageSize = pageSize;
+    this.genericClient = Objects.requireNonNull(genericClient, "genericClient");
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (nextActiveResponseIndex < activeResponse.getExecutionsCount()) {
+      return true;
+    }
+    fetch();
+    return nextActiveResponseIndex < activeResponse.getExecutionsCount();
+  }
+
+  @Override
+  public WorkflowExecutionInfo next() {
+    if (hasNext()) {
+      return activeResponse.getExecutions(nextActiveResponseIndex++);
+    } else {
+      throw new NoSuchElementException();
+    }
+  }
+
+  void fetch() {
+    if (nextResponse == null) {
+      // if nextResponse is null, it's the end of the iteration through the pages
+      return;
+    }
+
+    ListWorkflowExecutionsResponse response;
+    try {
+      response = this.nextResponse.get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e.getCause());
+    }
+
+    ByteString nextPageToken = response.getNextPageToken();
+    if (!nextPageToken.isEmpty()) {
+      ListWorkflowExecutionsRequest request =
+          ListWorkflowExecutionsRequest.newBuilder()
+              .setNamespace(namespace)
+              .setQuery(query)
+              .setPageSize(pageSize)
+              .setNextPageToken(nextPageToken)
+              .build();
+      this.nextResponse = genericClient.listWorkflowExecutionsAsync(request);
+    } else {
+      this.nextResponse = null;
+    }
+
+    if (response.getExecutionsCount() == 0 && nextResponse != null) {
+      log.warn(
+          "[BUG] listWorkflowExecutions received an empty collection with a non-empty nextPageToken");
+      // shouldn't be happening, but we want to tolerate it if it does, so we just effectively
+      // skip the empty response and wait for the next one in a blocking manner.
+      // If this actually ever happens as a normal scenario, this skipping should be reworked to
+      // be done asynchronously on a completion of nextResponse future.
+      fetch();
+      return;
+    }
+
+    activeResponse = response;
+    nextActiveResponseIndex = 0;
+  }
+
+  public void init() {
```

Comment:
As a public iterator, it can be hard to know that his must be called by a user. Wonder if we want to make the constructor and this package private. Or maybe just the whole class package private.

Reply:
The user shouldn't be creating this class, made package-private